### PR TITLE
add CatBoostPruningCallback

### PR DESCRIPTION
**Pull request by tohmae**
This PR addresses issue #2464.
If TrialPruned is raised at after_iteration in CatBoostPruningCallback, then catboost exits with error.
So I add check_pruned function in CatBoostPruningCallback.

I add an example code in the optuna-examples repository:

### CODE REVIEW

**Comment by nzw0301** [comment]
Brilliant! I'm looking forward to your updates. The commit history looks strange. Could you resolve it?

**Comment by tohmae** [comment]
Thank you for your comment.

If I merge master(#2618) in this branch, when I run sample code(for example [xgboost_simple.py](https://github.com/optuna/optuna-examples/blob/main/xgboost/xgboost_simple.py)), it alwayes ends with an error. But when I run with version 2.8.0, it succeeds. Isn't the same thing happening to you ?

```
$ python xgboost_simple.py 
[I 2021-06-09 21:07:35,240] A new study created in memory with name: no-name-5944bd96-07e8-4b2d-a695-4cce565a92d5
[I 2021-06-09 21:07:35,253] Trial 0 finished with value: 0.9020979020979021 and parameters: {'booster': 'gblinear', 'lambda': 0.00016458338949708223, 'alpha': 0.0005416674620586438, 'subsample': 0.3593187103440391, 'colsample_bytree': 0.696480150424724}. Best is trial 0 with value: 0.9020979020979021.
[I 2021-06-09 21:07:35,278] Trial 1 finished with value: 0.965034965034965 and parameters: {'booster': 'dart', 'lambda': 6.16568474831409e-06, 'alpha': 2.9323690893429236e-05, 'subsample': 0.6173849906062747, 'colsample_bytree': 0.8595021101176745, 'max_depth': 7, 'min_child_weight': 2, 'eta': 0.000569354042478708, 'gamma': 3.916019043978383e-08, 'grow_policy': 'lossguide', 'sample_type': 'weighted', 'normalize_type': 'tree', 'rate_drop': 1.033868829666294e-07, 'skip_drop': 0.0003122378294262642}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,293] Trial 2 finished with value: 0.8461538461538461 and parameters: {'booster': 'gblinear', 'lambda': 0.10242852791463153, 'alpha': 6.009539045659108e-06, 'subsample': 0.9978972253263705, 'colsample_bytree': 0.8424202225857023}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,314] Trial 3 finished with value: 0.8951048951048951 and parameters: {'booster': 'gbtree', 'lambda': 0.015913408455595322, 'alpha': 9.353404475640191e-06, 'subsample': 0.39785297147229515, 'colsample_bytree': 0.9980637973617716, 'max_depth': 5, 'min_child_weight': 9, 'eta': 2.31146277563236e-05, 'gamma': 0.02551692495071446, 'grow_policy': 'lossguide'}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,331] Trial 4 finished with value: 0.8951048951048951 and parameters: {'booster': 'gblinear', 'lambda': 1.0380673315585811e-07, 'alpha': 0.036560600416028585, 'subsample': 0.4370960620822139, 'colsample_bytree': 0.687332915041001}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,353] Trial 5 finished with value: 0.951048951048951 and parameters: {'booster': 'dart', 'lambda': 0.12501260037040057, 'alpha': 0.0002367460860547777, 'subsample': 0.892924032791395, 'colsample_bytree': 0.5081862582537645, 'max_depth': 7, 'min_child_weight': 5, 'eta': 0.06721232254422727, 'gamma': 3.3624178429133145e-06, 'grow_policy': 'depthwise', 'sample_type': 'uniform', 'normalize_type': 'forest', 'rate_drop': 0.03290617629836799, 'skip_drop': 0.000653977197349475}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,370] Trial 6 finished with value: 0.8811188811188811 and parameters: {'booster': 'gblinear', 'lambda': 2.640228148344214e-08, 'alpha': 0.005819709998269761, 'subsample': 0.24549706733652477, 'colsample_bytree': 0.7758331796515743}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,386] Trial 7 finished with value: 0.9230769230769231 and parameters: {'booster': 'gblinear', 'lambda': 0.0011791555384263394, 'alpha': 3.5313486244929844e-06, 'subsample': 0.4519940300289149, 'colsample_bytree': 0.7165206763934979}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,400] Trial 8 finished with value: 0.965034965034965 and parameters: {'booster': 'gblinear', 'lambda': 5.697259718576355e-07, 'alpha': 0.0001270313947222112, 'subsample': 0.8293021250646391, 'colsample_bytree': 0.8510821904544048}. Best is trial 1 with value: 0.965034965034965.
[I 2021-06-09 21:07:35,421] Trial 9 finished with value: 0.972027972027972 and parameters: {'booster': 'dart', 'lambda': 1.3612395598903652e-08, 'alpha': 0.0030153802836014, 'subsample': 0.6497731878729027, 'colsample_bytree': 0.4198710529357345, 'max_depth': 9, 'min_child_weight': 6, 'eta': 0.8951618267437211, 'gamma': 0.0016887642739510907, 'grow_policy': 'depthwise', 'sample_type': 'uniform', 'normalize_type': 'tree', 'rate_drop': 7.089540620091507e-08, 'skip_drop': 0.02247455178086285}. Best is trial 9 with value: 0.972027972027972.
[W 2021-06-09 21:07:35,462] Trial 10 failed because of the following error: IndexError('index -9223372036854775808 is out of bounds for axis 1 with size 2')
Traceback (most recent call last):
  File "/Users/t-ohmae/src/python/optuna/optuna/_optimize.py", line 216, in _run_trial
    value_or_values = func(trial)
  File "xgboost_simple.py", line 51, in objective
    param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
  File "/Users/t-ohmae/src/python/optuna/optuna/trial/_trial.py", line 505, in suggest_categorical
    return self._suggest(name, CategoricalDistribution(choices=choices))
  File "/Users/t-ohmae/src/python/optuna/optuna/trial/_trial.py", line 716, in _suggest
    param_value = self.study.sampler.sample_independent(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/sampler.py", line 401, in sample_independent
    mpe_above = _ParzenEstimator(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/parzen_estimator.py", line 83, in __init__
    categorical_weights = self._calculate_categorical_params(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/parzen_estimator.py", line 345, in _calculate_categorical_params
    weights[np.arange(n_observations), observations] += 1
IndexError: index -9223372036854775808 is out of bounds for axis 1 with size 2
Traceback (most recent call last):
  File "xgboost_simple.py", line 68, in <module>
    study.optimize(objective, n_trials=100, timeout=600)
  File "/Users/t-ohmae/src/python/optuna/optuna/study.py", line 401, in optimize
    _optimize(
  File "/Users/t-ohmae/src/python/optuna/optuna/_optimize.py", line 65, in _optimize
    _optimize_sequential(
  File "/Users/t-ohmae/src/python/optuna/optuna/_optimize.py", line 162, in _optimize_sequential
    trial = _run_trial(study, func, catch)
  File "/Users/t-ohmae/src/python/optuna/optuna/_optimize.py", line 267, in _run_trial
    raise func_err
  File "/Users/t-ohmae/src/python/optuna/optuna/_optimize.py", line 216, in _run_trial
    value_or_values = func(trial)
  File "xgboost_simple.py", line 51, in objective
    param["grow_policy"] = trial.suggest_categorical("grow_policy", ["depthwise", "lossguide"])
  File "/Users/t-ohmae/src/python/optuna/optuna/trial/_trial.py", line 505, in suggest_categorical
    return self._suggest(name, CategoricalDistribution(choices=choices))
  File "/Users/t-ohmae/src/python/optuna/optuna/trial/_trial.py", line 716, in _suggest
    param_value = self.study.sampler.sample_independent(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/sampler.py", line 401, in sample_independent
    mpe_above = _ParzenEstimator(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/parzen_estimator.py", line 83, in __init__
    categorical_weights = self._calculate_categorical_params(
  File "/Users/t-ohmae/src/python/optuna/optuna/samplers/_tpe/parzen_estimator.py", line 345, in _calculate_categorical_params
    weights[np.arange(n_observations), observations] += 1
IndexError: index -9223372036854775808 is out of bounds for axis 1 with size 2
```

**Comment by nzw0301** [comment]
@tohmae Thank you for sharing it. You are right. I could reproduce the bug on my local machine with the latest optuna as well.

**Comment by HideakiImamura** [comment]
Hi @tohmae. Thank you for reporting the fatal bug of the TPESampler. I fixed the error in https://github.com/optuna/optuna/pull/2739. Please wait to merge it to proceed with the discussion.

**Comment by nzw0301** [comment]
Thank you for updating the commit history! Could I start to review this PR?

**Comment by tohmae** [comment]
Thank you for your comments.
I would be grateful if you could review this PR.

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:53]: How about using `iteration` or `step`for cosnsistency?

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:78]: I understood that we cannot use `optuna.TrialPruned` in `after_iteration` of CatBoost. But it might cause some error for multi-thread training. I'll ask catboost developer to make it possible.

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Could you add the callback class into https://github.com/optuna/optuna/blob/master/optuna/integration/__init__.py?

---

Thank you for your pull request! As I mentioned in a comment of review, we should raise `TrialPruned` in `after_iteration`.  So I didn't check the test function yet. Let me ask the catboost developer team, so please give us a time.

File: optuna/integration/catboost.py
Comment (nzw0301) [reply optuna/integration/catboost.py:78]: https://github.com/catboost/catboost/issues/1744

**Comment by tohmae** [comment]
Thank you for your comments.
The PR has been modified to reflect the results of the review.

Thank you for making a feature request to catboost developer https://github.com/catboost/catboost/issues/1744.
But It will take a long time before the request is implemented, so I created an alternative PR.
If you have a time, could you comment ?

File: optuna/_optimize.py
Comment (nzw0301) [inline optuna/_optimize.py:240]: Unfortunately, this modification for catboost pruning might affect the existing code too much...

**Comment by nzw0301** [comment]
Thank you for your update! Unfortunately, introducing the logic to `optuna/_optimize.py` is a little bit too much change for only catboost pruning functionality.  So we need to wait for the response from the catboost team. I've added a comment to the issue on catboost repo.

**Comment by tohmae** [comment]
Thank you for your comments.
I don't think it is good to change the logic of _optimize.py at all.
I think it would be better if there is another method to integrate Optuna with external machine learning frameworks, especialy when callback function can not raise exception. (ex. when callback function executes in multithread.)
I want to wait for the response from the catboost team.

**Comment by HideakiImamura** [comment]
@tohmae @nzw0301 Any update on this?

**Comment by nzw0301** [comment]
No, because https://github.com/catboost/catboost/issues/1744 has not been updated.

**Comment by tohmae** [comment]
In order to prune with machine learning frameworks other than catboost, I add modify_result to Trial and call in _run_trial.
If you have a time, could you comment ?

**Comment by nzw0301** [comment]
Sorry for the delayed respose. I'm really suffering from bloody wrist pain. 

Similar to https://github.com/optuna/optuna/pull/2734#issuecomment-899014949, I'm afraid that such modification is too much for optuna's codebase to use catboost callback. One alternative way might be to implement this as an example instead of as optuna's callback. I'd like to hear dev member's opinion.

**Comment by nzw0301** [comment]
Awesome. I'll review this PR within the next week.

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:39]:
```suggestion
            If you set multiple datasets as ``eval_set``, the index of ``eval_set`` must be
            included in the ``valid_name``, e.g., ``validation_0`` or ``validation_1``
            when ``eval_set`` contains two datasets.
```

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [inline tests/integration_tests/test_catboost.py:30]:
```suggestion
    assert not pruning_callback.after_iteration(info)
```

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:35]:
```suggestion
            The name of the target validation dataset.
```

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:34]: If CatBoost always uses `validation` as a prefix, this part might be simplified. For example, it takes only `int`, whose default value is `0` and renames the argument name like `dataset_index` or something else from `valid_name`.

What do you think?

**Review by nzw0301 — CHANGES_REQUESTED** [review]
I'll give further comments within the next week, but let me share minor comments.

**Comment by nzw0301** [comment]
Sorry, one more thin. Could you add catboost related listing to https://github.com/optuna/optuna#integrations and https://github.com/optuna/optuna/blob/master/examples/README.md?

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:88]:
```suggestion
        """Raise :class:`optuna.TrialPruned` manually if the CatBoost optimization is pruned."""
```

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [inline tests/integration_tests/test_catboost.py:21]:
```suggestion
    assert pruning_callback.after_iteration(info)
```

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [inline tests/integration_tests/test_catboost.py:30]:
```suggestion
    assert not pruning_callback.after_iteration(info)
```

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [inline tests/integration_tests/test_catboost.py:33]: I think this test case is covered by `test_catboost_pruning_callback_init_param`. Can we remove this test?

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [inline tests/integration_tests/test_catboost.py:153]: In my understanding, `ValueError` is raised by the catboostcallback, so is `ValueError` more specific than `SystemError`?
```suggestion
    with pytest.raises(ValueError):
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Hi, sorry for leaving separate comments on this PR. Could you take a look at my comments?

File: tests/integration_tests/test_catboost.py
Comment (tohmae) [reply tests/integration_tests/test_catboost.py:121]: ValueError is raised by the catboostcallback, then SystemError is raised by catboost.
This is how it is currently implemented. Do you have a better idea ?

**Comment by tohmae** [comment]
Thank you for your comments.
The PR has been modified to reflect the results of the review except for one point.
> In my understanding, `ValueError` is raised by the catboostcallback, so is `ValueError` more specific than `SystemError`?

ValueError is raised by the catboostcallback, then SystemError is raised by catboost.
This is how it is currently implemented. Do you have a better idea ?

I would be grateful if you could review this PR.

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [reply tests/integration_tests/test_catboost.py:121]: Ah, indeed. Thank you for your explanation!

File: README.md
Comment (nzw0301) [inline README.md:103]: [memo]
The tree based link, https://github.com/optuna/optuna-examples/tree/main/catboost/catboost_pruning.py, is redirected to 
blob based link, https://github.com/optuna/optuna-examples/blob/main/catboost/catboost_pruning.py

File: optuna/integration/catboost.py
Comment (nzw0301) [reply optuna/integration/catboost.py:15]: Could you add this experimental decolator?

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:29]:
```suggestion
    .. note::
        :class:`optuna.TrialPruned` cannot be raised in :meth:`~optuna.integration.CatBoostPruningCallback.after_iteration` that is called in CatBoost via ``CatBoostPruningCallback``.
        You must call :meth:`~optuna.integration.CatBoostPruningCallback.check_pruned` after training manually unlike other pruning callbacks
        to raise :class:`optuna.TrialPruned`  mechanism of CatBoost.
```

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:48]:
```suggestion
    def __init__(self, trial: optuna.trial.Trial, metric: str, eval_set_index: Optional[int] = None) -> None:
```

File: optuna/integration/catboost.py
Comment (nzw0301) [inline optuna/integration/catboost.py:1]:
```suggestion
from typing import Any
from typing import Optional
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]

File: tests/integration_tests/test_catboost.py
Comment (HideakiImamura) [inline tests/integration_tests/test_catboost.py:121]: I'm not sure, but is the `ValueError` in `CatboostPruningCallback` inherited by `SystemError` which is raised by catboost?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for your continued contribution! Basically LGTM. I have several comments. Could you take a look?

File: tests/integration_tests/test_catboost.py
Comment (nzw0301) [reply tests/integration_tests/test_catboost.py:121]: https://github.com/optuna/optuna/pull/2734#issuecomment-1021260006

File: optuna/integration/catboost.py
Comment (nzw0301) [reply optuna/integration/catboost.py:107]: Hmm, good point. But we also would like to show the message like `" "Trial was pruned at iteration 10."` stored as `self._message`.

**Comment by nzw0301** [comment]
Sorry, let me confirm the state of this PR. Should we re-review this?

**Comment by tohmae** [comment]
Sorry for the late reply.
I have corrected everything except for the following review.

>How about returning a boolean value with respect to self._pruned=True or self._pruned=False? I think it would be more appropriate to check whether the trial has been pruned based on whether the return value of this method is True or not, and have the user raise TrialPruned, which is in line with the standard way of pruning in Optuna and the name of the method itself.

I have not decided how to correct the above points.
I would be grateful if you could advice.

File: optuna/integration/catboost.py
Comment (nzw0301) [reply optuna/integration/catboost.py:107]: Let me remind @HideakiImamura to discuss this comment.

Comment (HideakiImamura) [reply optuna/integration/catboost.py:107]: Thanks for the explanation. Understood. I think it is acceptable to raise `TrialPruned` in `check_pruned`. Let me resolve this conversation.

**Comment by HideakiImamura** [comment]
@tohmae @nzw0301 Sorry for the late reply. I checked the discussion, and understood why the current implementation raises `TrialPruned`. I agree with the current design to send the message of `"Trial was pruned at iteration {}.".format(step)`.

The code basically looks good to me. Let me start a final check.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the long running PR. LGTM!

**Comment by HideakiImamura** [comment]
@nzw0301 Please merge this PR after your approval.

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Could you take a look at my minor comment?

**Comment by tohmae** [comment]
Thank you for your comments.
I removed.

**Review by nzw0301 — APPROVED** [review]
Brilliant! Thank you for creating this long-running PR.

**Pull request merged**
